engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerService.java

```diff
@@ -0,0 +1,166 @@
+    private String analysisMount;
+    /**
+     * Socket to connect
+     */
+    @Value("${docker.host:#{null}}")
+    private String host;
+
+    @Value("${docker.image.default}")
+    private String defaultImage;
+
+    @Autowired
+    @Qualifier("analysisTaskExecutor")
+    private ThreadPoolTaskExecutor executor;
+
+    public DockerService(DockerRegistryProperties properties) {
+        DefaultDockerClientConfig.Builder builder = DefaultDockerClientConfig.createDefaultConfigBuilder()
+                .withDockerTlsVerify(false)
+                .withRegistryUrl(properties.getUrl())
+                .withRegistryPassword(properties.getPasword())
+                .withRegistryUsername(properties.getUsername());
+        Optional.ofNullable(host).ifPresent(builder::withDockerHost);
+        DefaultDockerClientConfig config = builder.build();
+
+        URI host = config.getDockerHost();
+        DockerHttpClient dockerHttpClient = new ApacheDockerHttpClient.Builder()
+                .dockerHost(host)
+                .sslConfig(config.getSSLConfig())
+                .maxConnections(50)
+                .build();
+        client = DockerClientImpl.getInstance(config, dockerHttpClient);
+        log.info("Initialized docker interface [{}]", host);
+    }
+
+    @Override
+    protected Overseer analyze(
+            AnalysisSyncRequestDTO analysis, File analysisDir, BiConsumer<String, String> sendCallback, Integer updateInterval, KrbConfig krbConfig
+    ) {
+        List<String> env = buildRuntimeEnvVariables(analysis.getDataSource(), krbConfig.getIsolatedRuntimeEnvs()).entrySet().stream().map(entry ->
+                entry.getKey() + "=" + entry.getValue()).collect(Collectors.toList()
+        );
+        Instant started = Instant.now();
+        Long id = analysis.getId();
+        String image = Optional.ofNullable(analysis.getDockerImage()).orElse(defaultImage);
+
+        StringBuffer stdout = new StringBuffer();
+        String script = WORKDIR + "/" + analysis.getExecutableFileName();
+        BiConsumer<String, String> callback = (stage, out) -> {
+            stdout.append(out);
+            sendCallback.accept(stage, out);
+        };
+
+        CompletableFuture<String> init = CompletableFuture.supplyAsync(
+                () -> {
+                    log.info("Execution [{}] use docker image [{}]", id, image);
+                    pullImage(callback, id, image, stdout, client);
+                    callback.accept(Stage.INITIALIZE, "Pull complete, creating container\n");
+                    log.info("Execution [{}] creating container with image [{}]", id, image);
+                    String containerId = createContainer(analysisDir, env, image, script);
+                    log.info("Execution [{}] created container [{}]", id, containerId);
+                    callback.accept(Stage.INITIALIZE, "Container [" + containerId + "] created with [" + analysisDir.getPath() + "] mounted for execution\n");
```

Comment:
The concept of stage is a bit different: at each stage, the process is either progressing or in error.
If there is error, further change is unlikely (but not completely ruled out, because racing condition with aborting might occur). 
If there is no error, the process will eventually progress to the next stage (except for COMPLETED stage) or fall into error.
This allows to avoid the need to multiply the states to reflect error conditions and ensures that if somebody adds a new stage later on, well-written consumers can handle that by default, as in most cases they will use either error condition or specific stage name to match in their conditions.